### ISSUE TITLE
Remove `distinguish_since` capability

### ISSUE THREAD

**Issue by frankmcsherry** [frankmcsherry/differential-dataflow#90]
Currently any `TraceAgent`, the thing you hold on to to be able to re-use an arrangement, has two capabilities (with horrible names):

1. `advance_by`: indicates the lower bound (frontier) on times that maybe be used as comparisons against the entries in the trace. As this advances, the trace is able to compact its own representation by determining that some times are equivalent, and their differences coalesced.

2. `distinguish_since`: independently, each trace maintains boundaries between each of the batches it has created, and waits for trace agents to decline an interest in keeping them from merging. This is so that operators like `join` can get access to what may be historical traces, because they `advance_by` a trace based on the progress in the other input (weird, but true).

The times in the `distinguish_since` capability currently have no particular semantic use, and could as easily be replaced with the sequence number of the batch. Although one could imagine a general and rich "bookmarking" functionality, the current use is "for this batch, could I have the trace up to but not including it, please".

Rather than track two separate capabilities, which causes a fair amount of headache (especially when they head in different directions), I propose that we scrap the `distinguish_since` capability, and simply capture the state of the trace at the moment each batch is created (in the form of a list of batches). The associated streams of data would change from "batch" to "batch + history", and allow each implementation to hold on to the history as long as appropriate, without blocking other users of the trace.

One possible downside of this is that the `distinguish_since` capability was used as a way of blocking merging until a user had downgraded their `advance_by` capabilities, increasing the chance that when time came for a merge there would be the potential to collapse the updates as well. Now as soon as a batch is entered into the trace it can be merged, even if it would be best overall for that merging to be delayed just a moment. Perhaps we can build in a one-call delay to give each downstream operator a chance to perform the associated work.

This might also be a good time to rename the `advance_by` capability, which describes what the trace does to each of the times (they are "advanced by" the frontier), but isn't especially descriptive for the user of the trace. Awkwardly, it seems that "distinguish since" is a lot closer to what the capability represents (all times in advance of the frontier must be distinguishable from each other). I'd prefer not to simply swap that name in, in the interest of shaking out bugs where we forget to rename things, but it might be a good time to float some synonyms.

**Comment by utaal** [comment]
If we were to keep them both around, what about `merge_batches_until` and `distinguish_(timestamps_)since`?

Other options for `advance_by`:
* `keep_history_since`;
* `advance_timestamps_until/to`.

**Comment by frankmcsherry** [comment]
I started on these edits, and one complication shows up (at least one):

It is currently possible to have an `Arranged` source of data, meaning stream of batches and indexed trace, where all of the trace handles have been dropped and so the trace is no longer maintained. An example of this is in the `join` operator which wants each input arranged, but only needs to maintain the history if the opposite stream is still active. If an opposite stream is terminated (corresponding to now immutable data), the first stream can cease its trace maintenance.

This worked in part because access to the trace was mediated through the trace handle, and by dropping it there was no other route for trace data to arrive. If instead we ship `(Batch, Vec<Batch>)` data along channels, ... we need to be clear about what this all means.

There is nothing fundamentally hard about doing this, in that we can use the same signals (trace handle drop) to clean up the trace, at which point we have no batches to fill up a `Vec<Batch>` which .. is fine as long as the recipient understands that having dropped the trace handle these are no longer valid history (i.e., an empty list does not indicate an empty history, just an absent history).

This doesn't seem hard to address "by convention", but it does signal that we are probably planning on accessing trace data without using the help of the trace. We could instead ship `(Batch, Option<Trace::Cursor::Storage>)`, which would still require the trace's help to get a cursor, and only allow navigation if we are both received a `Some(_)` variant and have a trace handle from which to grab a cursor.

**Comment by utaal** [comment]
To have time to serialise things (unless this is done at batch construction time, somehow), it may be helpful to keep the batches in advance of (the old) `advance_by` from being merged (while they're being written down): this way one could trivially write down all the new batches one by one, and just load them all as non-merged batches on recovery.

I may need a clarification: with the current (and proposed) interface, merging two batches `a` and `b` may result in a merged batch `c` with `c.since > a.since` and `c.since > b.since`?

**Comment by frankmcsherry** [comment]
I was guessing that serialization would probably be done at batch creation time, but I see your point if not. I have some other reservations about the `distinguish_since` API *and* the proposed fix, so let's think out what we would want from it.

I like a "bookmark" sort of feature, where you can hold on to a capability for a batch boundary (e.g. a `lower` or an `upper`) that prevents merging across that boundary. I think that this does something like what serialization would want (e.g. "I've serialized up to `upper`; please don't merge across that, but feel free to merge among batches after it"). Does that sound believable to you?

**Comment by utaal** [comment]
> I like a "bookmark" sort of feature, where you can hold on to a capability for a batch boundary (e.g. a lower or an upper) that prevents merging across that boundary. I think that this does something like what serialization would want (e.g. "I've serialized up to upper; please don't merge across that, but feel free to merge among batches after it"). Does that sound believable to you?

Would this also ensure that `since` ≤ `upper`? I.e. prevent merging *and* timestamp advancement across that boundary?

**Comment by frankmcsherry** [comment]
> I may need a clarification: with the current (and proposed) interface, merging two batches a and b may result in a merged batch c with c.since > a.since and c.since > b.since?

I think that when you merge two batches, you need to advance the `since` bound to not make any promises that the merged data cannot provide. What probably happens at the moment (and the code tests for) is that one of `a.since` and `b.since` should be in advance of the other (because the trace's frontier used for them only advances), and we pick the more advanced one. Otherwise, we would need to pick a frontier in advance of both of them.

**Comment by frankmcsherry** [comment]
> Would this also ensure that since ≤ upper? I.e. prevent merging and timestamp advancement across that boundary?

No, that was not the intent. If you want to prevent advancement, you can indicate that with a trace handle by not advancing its `advance_by` (gah, the names...).

Perhaps I should try and clear up an unspoken desideratum, which is that holding these capabilities shouldn't cripple the other users of the trace by preventing e.g. merging, but perhaps also advancement. Most of the requirements have been historical ("I need access to the history, with some properties"), whereas the durability work seems like it might be a future constraint ("I will need history from here held separate and not advanced").

The poster-child for bad behavior at the moment is in the `server` project, where the trace handle for the random graph stream downgrades none of its capabilities, preventing merging and advancement. I can understand not downgrading advancement, if you want full historical detail, but not downgrading merging seems like just a performance loss there. On the other hand, we do want to be able to hook the batches for durability somehow.

**Comment by utaal** [comment]
> whereas the durability work seems like it might be a future constraint. I will need history from here held separate and not advanced.

Sounds right to me, right now, but there's a chance this can be relaxed.

> I can understand not downgrading advancement, if you want full historical detail, but not downgrading merging seems like just a performance loss there.

I see, makes sense.
